ROX:14415: Update docs with port numbers for connections (#80322)
openshift/openshift-docs: modules/acs-architecture-connections.adoc

.{product-title-short} connections between components

[cols="4",options="header"]
|===
| Component or external entity | Connection type | Port | Additional information

| Central and Sensor on secured cluster
a| * TCP/HTTPS
* gRPC
| 443
| Sensor and Central primarily communicate over a bidirectional gRPC stream, initiated by Sensor to Central's port 443.

| Central and user (CLI)
a| * gRPC
* HTTPS (with `--force-http1` option)
| 443
| For more information on the `--force-http1` option, see the `roxctl` command options in "roxctl CLI command reference".

| Central and vulnerability feeds
| HTTPS
| 443

> | Connects to definitions.stackrox.io` by default.

Reviewer: Missing backtick?

> | Scanner to Sensor

Reviewer: This connection happens within the secured cluster - not relevant for the firewall.

| HTTPS
| 443

> | Only in secured clusters.

Reviewer: The central-scanner never reaches directly to Sensor. I think, we may need to use a similar comment as in the row below: `For delegated scanning scenarios or in {ocp} secured clusters.`

// Module included in the following assemblies:
//
// * architecture/acs-architecture.adoc
:_mod-docs-content-type: CONCEPT
[id="acs-architecture-connections_{context}"]
= {product-title-short} connection protocols and default ports

The protocols and default ports for some connections within {product-title-short} and between {product-title-short} and external components are shown in the following table. You might need this information for configuring the firewalls in your system to allow traffic on certain ports.

.{product-title-short} connections between components

[cols="4",options="header"]
|===
| Component or external entity | Connection type | Port | Additional information

> | Central and sensor on secured cluster

Reviewer 2: ```suggestion
| Central and Sensor on secured cluster
```

a| * TCP/HTTPS
* gRPC
| 443

> | Sensor and Central primarily communicate over a bidirectional gRPC stream, initiated by sensor to Central's port 443.

Reviewer 2: ```suggestion
| Sensor and Central primarily communicate over a bidirectional gRPC stream, initiated by Sensor to Central's port 443.
```

> | Central and user ({product-title-short} web portal or REST API)

Reviewer 3: I still think for the non-bidirectional cases, it's unclear which component the port is associated with. Can we update the table or these titles to make it clear that "The user reaches out to Central on port 443 via web portal or REST API" or later "Scanner reaches out to Scanner DB at port 5432"

| HTTPS or gRPC
| 443
|

| Central and user (CLI)
| gRPC
| 443
|

| Central and vulnerability feeds
| HTTPS
| 443
|

> | Scanner and Scanner-DB

Reviewer 3: Do we want to include Scanner V4 components, too? If so there are a few more:

Scanner V4 Indexer <-> Central (depends on the context)
Scanner V4 Indexer -> Scanner V4 DB (TCP 5432)
Scanner V4 Matcher <-> Central (depends on the context)
Scanner V4 Matcher -> Scanner V4 Indexer (gRPC I forget the port, but I can get it, if you want it)
Scanner V4 Matcher -> Scanner V4 DB (5432 TCP)

| TCP
| 5432
|

| Sensor and Admission Controller
| gRPC
| 443
| This is a bidirectional gRPC stream, initiated by Admission Controller to Sensor's port 443.

> | Sensor and collector

Reviewer 2: ```suggestion
| Sensor and Collector
```

| gRPC
| 443

> | This connection is bidirectional and initiated by collector to Central's port 443. If node scanning is enabled for {ocp}, a connection on port 8444 is initiated by sensor to compliance (running in the collector pod).

Reviewer: ```suggestion
| This connection is bidirectional and initiated by collector to Sensor's port 443. If node scanning is enabled for {ocp}, a connection on port 8444 is initiated by Sensor to compliance (running in the collector pod).
```

Reviewer: Regarding bidirectionally: it will be bidirectional in 4.6, now it is mainly Collector -> Sensor.
Sensor can however communicate with Compliance container (living inside the Collector pod), so I guess this is bidirectional in 4.5 and earlier.

> | Sensor and admission controller
> | gRPC
> | 443
> | This is a bidirectional gRPC stream, initiated by the admission controller to sensor's port 443.
> 
> 
> | Sensor and collector
> | gRPC
> | 443
> | This connection is bidirectional and initiated by collector to Central's port 443. If node scanning is enabled for {ocp}, a connection on port 8444 is initiated by sensor to compliance (running in the collector pod).

Reviewer 2: ```suggestion
| Sensor and Admission Controller
| gRPC
| 443
| This is a bidirectional gRPC stream, initiated by the Admission Controller to Sensor's port 443.

| Sensor and collector
| gRPC
| 443
| This connection is bidirectional and initiated by collector to Central's port 443. If node scanning is enabled for {ocp}, a connection on port 8444 is initiated by Sensor to compliance (running in the Collector pod).
```
From my understanding, and please correct me if I'm wrong, `Admission Controller`, `Sensor` and `Collector` are all proper nouns that should be capitalized. Unfortunately searching through our repo found many entries of each both capitalized and not, I think we should make it all consistent in the future.

Author: Correct. For Red Hat components, and Kubernetes components, we capitalize per IBM and [Red Hat Style](https://redhat-documentation.github.io/supplementary-style-guide/#operator). We are trying to be consistent going forward.

// Module included in the following assemblies:
//
// * architecture/acs-architecture.adoc
:_mod-docs-content-type: CONCEPT
[id="acs-architecture-connections_{context}"]
= {product-title-short} connection protocols and default ports

The protocols and default ports for connections within {product-title-short} and between {product-title-short} and external components are shown in the following table.

.{product-title-short} connections between components

[cols="3",options="header"]
|===

> | Component or external entity | Connection type | Port

Reviewer 3: I think the port only makes sense when talking about the direction of the request as well. For example, for `Scanner and Scanner-DB | TCP | 5432`, this only makes sense when it's clear that Scanner reaches out to Scanner DB at port 5432.

At the moment, the table makes it seem like the communication may go either way and use that same port for both directions.

> | Central and external services, for example, Jira and PagerDuty

Reviewer 4: +1 
If we are talking about external integrations and callbacks, these are custom in each customers environment.
There will likely be no standard ports we could really require here.

| HTTPS
| ?

> | Central and secured cluster

Reviewer 4: The primary communication between Central and the Cluster happens through Sensor.
Sensor could be seen as the proxy for everything that is running in a Secured Cluster.
I don't know whether we should split out Central <-> Secured Cluster comms in here.

| HTTPS
| 443

| Central and sensor
| TCP

> | 443

Reviewer 3: are these different?

Author: Actually, I guess TCP is the physical and HTTP is the protocol on top, and they don't need to be broken out?

Reviewer 3: Inside the Secured Cluster, only two components reach out to Central:

* Sensor - has a persistent gRPC connection to Central. Port is 443 (I think)
* Admission Controller - rare, but it is sometimes allowed to reach out to Central, directly. Probably HTTP 443, but I actually don't know

Reviewer 4: We should probably differentiate between HTTP(S) and gRPC. 
The latter can be filtered out by Layer 7 Firewalls / WAFs, so it would need to be explicitly allowed in such environments.
Primarily, Sensor and Central communicate over a bidirectional gRPC stream, initiated by Sensor to Central's port 443.

> | Central and user ({product-title-short} web portal or REST API)

Reviewer 3: Users may use HTTPS, but I believe gRPC is allowed, too, if the user wanted to (at the same port)

| HTTPS
| 443

| Central and user (CLI)

> | gRPC

Reviewer 3: this is mostly true, though I believe there is a setting which forces HTTPS traffic (probably uncommonly used, so may not be worth mentioning)

Reviewer 3: I still think it's worth checking with the Merlin team(?) to see if we should mention HTTPS here, too

Reviewer 5: @kcarmichael08 As I can see from `roxctl --help` - we have:
```
     --force-http1                Always use HTTP/1 for all connections (advanced; only use if you encounter connection issues). Alternatively, enable by setting the ROX_CLIENT_FORCE_HTTP1 environment variable to true (default false)
```
This ticket: https://issues.redhat.com/browse/ROX-5787
And changelog for v51: https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md#510

Maybe you can change it in doc, similar to `Central and Sensor on secured cluster` section:
- gRPC
- HTTPS (with `--force-http1` option)

| 443

| Scanner and Scanner-DB
| TCP
| 5432

| Scanner and vulnerability feeds
| HTTPS

> | ?

Reviewer 3: Scanner does not reach out to the vuln data. Instead it's Central that does it. And it's probably 443 (default HTTPS port)

| Scanner and image registries
| HTTPS

> | ?

Reviewer 3: depends on the registry. I don't think this is worth mentioning

| Sensor and admission controller
| gRPC

> | ?

Reviewer 4: This is a bidirectional gRPC stream, initiated by Admission Controller to sensor's port 443

| Sensor and collector
| gRPC

> | ?

Reviewer 4: Bidirectional gRPC connection initiated by Collector to sensor's port 443.
If Node Scanning is enabled (OpenShift 4), additionally port 8444, initiated by Sensor to compliance (running in Collector pod)

|===

> //old graphic version: https://docs.openshift.com/acs/3.70/architecture/acs-architecture.html#external-components_acs-architecture

Reviewer 3: curious: why switch from graphic to tabular view?

Author: I wasn't involved in that decision but I think because it takes months to get graphics updated and in theory, a text update is quicker, if info changes.

// Module included in the following assemblies:
//
// * architecture/acs-architecture.adoc
:_mod-docs-content-type: CONCEPT
[id="acs-architecture-connections_{context}"]
= {product-title-short} connection protocols and default ports

Components of {product-title-short} use various default ports and connection protocols. Depending on your system and firewall configuration, you might need to configure your firewall to allow traffic on certain ports.

> The following table provides default ports and protocols for some connections within {product-title-short} and between {product-title-short} and external components. This is helpful for configuring your firewall to allow traffic to and from the cluster. However, you might need more detailed information in some scenarios. For example, if your firewall is integrated in the cluster router, you might need to specify ports for connections that happen within one cluster but might be on different IP networks. In this scenario, you can use the {product-title-short} network policy YAML files in your {ocp} and Kubernetes clusters to determine connections and ports that you might need to configure.

Reviewer 6: Suggestion, totally up to you:
```suggestion
The following table provides default ports and protocols for some connections within {product-title-short} and between {product-title-short} and external components. This is helpful for configuring your firewall to allow inbound and outbound cluster traffic. However, you might need more detailed information in some scenarios. For example, if your firewall is integrated in the cluster router, you might need to specify ports for connections that happen within one cluster but might be on different IP networks. In this scenario, you can use the {product-title-short} network policy YAML files in your {ocp} and Kubernetes clusters to determine connections and ports that you might need to configure.
```

> The following table provides default ports and protocols for some connections within {product-title-short} and between {product-title-short} and external components. This is helpful for configuring your firewall to allow traffic to and from the cluster. However, you might need more detailed information in some scenarios. For example, if your firewall is integrated in the cluster router, you might need to specify ports for connections that happen within one cluster but might be on different IP networks. In this scenario, you can use the {product-title-short} network policy YAML files in your {ocp} and Kubernetes clusters to determine connections and ports that you might need to configure.

Reviewer 6: Just a thought, would it be possible to split this paragraph into two somewhere? Not a big deal, I am just personally afraid of any "wall of text" effect reducing the users' willingness to read.

Maybe like this?
```suggestion
The following table provides default ports and protocols for some connections within {product-title-short} and between {product-title-short} and external components. This is helpful for configuring your firewall to allow traffic to and from the cluster. 

However, you might need more detailed information in some scenarios. For example, if your firewall is integrated in the cluster router, you might need to specify ports for connections that happen within one cluster but might be on different IP networks. In this scenario, you can use the {product-title-short} network policy YAML files in your {ocp} and Kubernetes clusters to determine connections and ports that you might need to configure.
```

.{product-title-short} connections between components

[cols="4",options="header"]
|===
| Component or external entity | Connection type | Port | Additional information

| Central and Scanner V4 Indexer
| gRPC
| 8443
|

| Central and Sensor on secured cluster
a| * TCP/HTTPS
* gRPC
| 443
| Sensor and Central primarily communicate over a bidirectional gRPC stream, initiated by Sensor to Central's port 443.

| Central and user (CLI)
a| * gRPC
* HTTPS (with `--force-http1` option)
| 443

> | For more information on the `--force-http1` option, see the `roxctl` command options in "roxctl CLI command reference".

Reviewer 6: If you do end up changing the link text of that additional resources link, this should probably be updated as well.

[id="acs-architecture-connections_{context}"]
= {product-title-short} connection protocols and default ports

The protocols and default ports for some connections within {product-title-short} and between {product-title-short} and external components are shown in the following table. You might need this information for configuring the firewalls in your system to allow traffic on certain ports.

.{product-title-short} connections between components

[cols="4",options="header"]
|===
| Component or external entity | Connection type | Port | Additional information

| Central and Sensor on secured cluster
a| * TCP/HTTPS
* gRPC
| 443
| Sensor and Central primarily communicate over a bidirectional gRPC stream, initiated by Sensor to Central's port 443.

| Central and user ({product-title-short} web portal or REST API)
| HTTPS or gRPC
| 443
| The user reaches out to Central on port 443 by using the web portal or the REST API.

| Central and user (CLI)
| gRPC
| 443
|

| Central and vulnerability feeds
| HTTPS
| 443
|

| Scanner to Scanner-DB
| TCP
| 5432
| Scanner reaches out to Scanner DB at port 5432.

| Scanner V4 Indexer and Central
|
|
| Protocol and port used depends on the context.

> Scanner V4 Indexer to Scanner V4 DB

Reviewer 5: I have noticed on the rendered page broken formatting. Maybe this line causes problems.
```suggestion
| Scanner V4 Indexer to Scanner V4 DB
```

Besides that. 
- Since we are already listing `Scanner -> Scanner DB` connection, we should also add `Central -> Central DB`.
- Also `Sensor -> Scanner` connection is not mentioned here.

Reviewer 5: @kcarmichael08 - it should be like this:
- `Central -> Central DB` - should be the same as `Scanner -> Scanner DB` (TCP, 5432). btw. This can be different if a customer uses "bring your own DB". https://docs.openshift.com/acs/4.5/installing/installing_ocp/install-central-ocp.html#install-central-operator-external-db_install-central-ocp

- `Sensor -> Scanner` - should be the same as `Central -> Scanner` (maybe this can be grouped together). Probably @RTann knows better how `Sensor` "talks" with `Scanner`.

// Module included in the following assemblies:
//
// * architecture/acs-architecture.adoc
:_mod-docs-content-type: CONCEPT
[id="acs-architecture-connections_{context}"]
= {product-title-short} connection protocols and default ports

The protocols and default ports for some connections within {product-title-short} and between {product-title-short} and external components are shown in the following table. You might need this information for configuring the firewalls in your system to allow traffic on certain ports.

.{product-title-short} connections between components

[cols="4",options="header"]
|===
| Component or external entity | Connection type | Port | Additional information

| Central and Central DB
| TCP
| 5432
| If you are using an external database with Central, the connection type and port might differ. See "Installing Central with an external database using the Operator method" for more information.

| Central and Sensor on secured cluster
a| * TCP/HTTPS
* gRPC
| 443
| Sensor and Central primarily communicate over a bidirectional gRPC stream, initiated by Sensor to Central's port 443.

> | Central and user ({product-title-short} web portal or REST API)
> | HTTPS or gRPC
> | 443
> | The user reaches out to Central on port 443 by using the web portal or the REST API.

Reviewer 4: ```suggestion
```
Purely internal

| Central and user (CLI)
a| * gRPC
* HTTPS (with `--force-http1` option)
| 443
| For more information on the `--force-http1` option, see the `roxctl` command options in "roxctl CLI command reference".

| Central and vulnerability feeds
| HTTPS
| 443

> |

Reviewer 3: Maybe mention this is definitions.stackrox.io, by default

> |

Reviewer 4: ```suggestion
| Connects to `definitions.stackrox.io` by default.
```

> | Scanner to Scanner DB
> | TCP
> | 5432
> | Scanner reaches out to Scanner DB at port 5432.

Reviewer 4: ```suggestion
```
Internal only - except if a user brings their own database, in which case I'd expect them to know on which port that DB is available.

| Scanner V4 Indexer and Central

> |

Reviewer 3: Indexer -> Central 443 HTTPS
Central -> Indexer 8443 gRPC

Reviewer 4: @RTann  Do we have cases where Central and the Indexer run on different clusters? 
If not, we can remove this entry as well

|
| Protocol and port used depends on the context.

> | Scanner V4 Indexer to Scanner V4 DB
> | TCP
> | 5432

Reviewer 4: ```suggestion
```
Internal only

Reviewer: Better keep it - Service IPs may be different L3 network.

|

| Scanner V4 Matcher and Central (depends on the context)

> |

Reviewer 3: Same idea as Indexer and Central

Reviewer 4: @RTann Same as above: Can we encounter situations where Central and the Matcher run in different clusters?
If not, this can also be removed from this overview

|
| Protocol and port used depends on the context.

> | Scanner V4 Matcher to Scanner V4 Indexer
> | gRPC
> |
> |

Reviewer 4: ```suggestion
```
Internal only

> | Scanner V4 Matcher to Scanner V4 DB

Reviewer 3: There is also Matcher -> Indexer which is gRPC 8443

> | Scanner V4 Matcher to Scanner V4 DB
> | TCP
> | 5432
> |

Reviewer 4: ```suggestion
```
Internal only

Reviewer: Better leave it - Service IPs may be different L3 network.

| Sensor and Admission Controller
| gRPC
| 443
| This is a bidirectional gRPC stream, initiated by Admission Controller to Sensor's port 443.

> | Sensor and Collector
> | gRPC
> | 443
> | This connection is bidirectional. Collector initiates a connection to Sensor on port 443. If node scanning is enabled for {ocp}, a connection on port 8444 is initiated by Sensor to the Compliance container located in the Collector pod.
> 

Reviewer 4: ```suggestion
```
This is a bit of a mixed bag. 
**IF** the Node running a Collector DaemonSet is located in a different network than the Node running Sensor, there might be firewalls crossed - while possible, I think this scenario is unlikely to happen.
This would be a good candidate for a detail page (e.g. "Secured Cluster Internals").

Reviewer: Sensor is accessed over Service IP which can be in a different network (in terms of L3 IP network).

> | Sensor to Scanner

Reviewer 3: This is only for delegated scanning scenarios or in OpenShift Secured Clusters

Sensor -> Scanner - 8443 gRPC
Scanner -> Sensor (only in secured clusters) - 443 HTTPS

Reviewer 3: There is also related Scanner V4 Indexer and Sensor in Secured Clusters (and/or delegated scanning scenarios)
